contrib/babelfishpg_common/src/datetime.c

```diff
@@ -679,6 +682,596 @@ datetime_pl_datetime(PG_FUNCTION_ARGS)
+			if (!isTextMonthPresent(field[i]))
+				return false;
+			count_string++;
+		}
+		else
+			return false;
+	}
+	if (count_number == 2 && count_string == 1)
+	{
+		/*
+		 * If the 1st field or 3rd field is a string then swap with the 2nd field
+		 * as when the date is given separatly then all different forms of
+		 * dates is supported. To avoid the conversion failure in `isTextMonthPresent`
+		 * later we are swapping earlier. And convert to Supported date formats which are
+		 * {"DD-MON-YYYY", "YYYY-MON-DD"}
+		 *
+		 * For example convert the input from "JULY 23 2000" to "23 JULY 2000"
+		 * and convert the input from "23 2000 JULY" to "23 JULY 2000".
+		 */
+		if (ftype[0] == DTK_STRING)
+		{
+			SWAP_FIELDS(0, 1);
+		}
+		else if (ftype[2] == DTK_STRING)
+		{
+			SWAP_FIELDS(1, 2);
+		}
+		return true;
+	}
+	return false;
+}
+
+/*
+ * pltsql_time_in
+ *		The function modifies the given input into valid form and
+ *		stores the output to result. An error will be thrown if
+ *		not an valid form.
+ * Returns
+ * 		false - If the sql_dialect is not an TSQL and proceed with
+ * 			the original behaviour
+ * 		true - If the sql_dialect is an TSQL and successfully converted
+ * 			the input to time datatype
+ */
+bool pltsql_time_in(const char* str, int32 typmod, TimeADT *result)
+{
+	fsec_t fsec;
+	struct pg_tm tt, *tm = &tt;
+	int tz, nf, dterr, dtype, ftype[MAXDATEFIELDS];
+	char workbuf[MAXDATELEN + 1], *field[MAXDATEFIELDS];
+	StringInfo res;
+	regex_t time_regex;
+	char *pattern, *temp_field;
+	int len_str = strlen(str);
+	char *modified_str = (char*) palloc(len_str + 1);
+	int j = 0, i = 0;
+
+	/* If sql_dialect is not an TSQL then return false */
+	if (sql_dialect != SQL_DIALECT_TSQL)
+		return false;
+
+	/* Throw a common error message while casting to time datatype */
+	#define TIME_IN_ERROR()	\
+		ereport(ERROR,	\
+				(errcode(ERRCODE_INVALID_DATETIME_FORMAT),	\
+				 errmsg("Conversion failed when converting date and/or time from character string.")));	\
+
+	/*
+	 * The below logic removes the space present before and after of ':', '/',
+	 * '.', '-' and also removes the multiple whitespaces to single whitespace.
+	 * For example converts the input from " 2023-  12 - 12   1  :     1" to "2023-12-12 1:1"
+	 */
+	while (i < len_str)
+	{
+		if (str[i] == ' ' && (i == (len_str - 1) || str[i+1] == ' ' || str[i+1] == ':' || str[i+1] == '/' || str[i+1] == '.' || str[i+1] == '-'))
+		{
+			i++;
+			continue;
+		}
+		else if (str[i] == ':' || str[i] == '/' || str[i] == '.' || str[i] == '-')
+		{
+			modified_str[j++] = str[i++];
+			while (i < len_str && str[i] == ' ')
+				i++;
+		}
+		else
+			modified_str[j++] = str[i++];
+	}
+	modified_str[j] = '\0';
+
+	dterr = ParseDateTime(modified_str, workbuf, sizeof(workbuf),
+			field, ftype, MAXDATEFIELDS, &nf);
+
+	/*
+	 * If there are no errors while parsing the modified_str then
+	 * validate the fields based on their types.
+	 */
+	if (dterr == 0)
```

Comment:
this is a big if block, can we break relevant functionality into separate smaller functions